mne/beamformer/_rap_music.py

```diff
@@ -0,0 +1,253 @@
+
+                source_pos = forward['source_rr'][i_source]
+                if n_orient == 1:
+                    source_ori = forward['source_nn'][i_source]
+
+        idx_k = slice(n_orient * source_idx, n_orient * (source_idx + 1))
+        Ak = G[:, idx_k]
+        if n_orient == 3:
+            Ak = np.dot(Ak, source_ori)
+
+        A[:, k] = Ak.ravel()
+
+        if return_explained_data:
+            gain_k = gain[:, idx_k]
+            if n_orient == 3:
+                gain_k = np.dot(gain_k, source_ori)
+            gain_dip[:, k] = gain_k.ravel()
+
+        oris[k] = source_ori
+        poss[k] = source_pos
+        if n_orient == 3:
+            oris[k] = np.dot(forward['source_nn'][idx_k], oris[k])
+
+        logger.info("source %s found: p = %s" % (k + 1, source_idx))
+        if n_orient == 3:
+            logger.info("ori = %s %s %s" % tuple(oris[k]))
+
+        projection = _compute_proj(A[:, :k + 1])
+        G_proj = np.dot(projection, G)
+        phi_sig_proj = np.dot(projection, phi_sig)
+
+    sol = linalg.lstsq(A, data)[0]
+
+    explained_data = None
+    if return_explained_data:
+        explained_data = np.dot(gain_dip, sol)
+
+    return _make_dipoles(times, forward['src'], poss,
+                         oris, sol), explained_data
+
+
+def _make_dipoles(times, src, poss, oris, sol):
+    amplitude = sol * 1e9
+    oris = np.array(oris)
+    gof = []
+
+    dipoles = []
+    for i_dip in range(poss.shape[0]):
+        i_pos = poss[i_dip][np.newaxis, :].repeat(len(times), axis=0)
+        i_ori = oris[i_dip][np.newaxis, :].repeat(len(times), axis=0)
+        dipoles.append(Dipole(times * 1e3, i_pos, amplitude[i_dip],
+                              i_ori, gof))
+
+    return dipoles
+
+
+def _compute_subcorr(G, phi_sig):
+    """ Compute the subspace correlation
+    """
+    Ug = linalg.qr(G, mode='economic')[0]
+    tmp = np.dot(Ug.T.conjugate(), phi_sig)
+    subcorr = np.dot(tmp, tmp.T.conjugate()).real
+    eig_vals, eig_vecs = linalg.eigh(subcorr)
+    return np.sqrt(eig_vals[-1]), eig_vecs[:, -1]
+
+
+def _compute_proj(A):
+    """ Compute the orthogonal projection operation for
+    a manifold vector A.
+    """
+    Ah = A.T.conjugate()
+    I = np.identity(A.shape[0])
+    return I - np.dot(np.dot(A, linalg.pinv(np.dot(Ah, A))), Ah)
+
+
+@verbose
+def rap_music(evoked, forward, noise_cov, signal_ndim=None, n_dipoles=5,
+              return_residual=False, picks=None, verbose=None):
+    """RAP-MUSIC source localization method.
+
+    Compute Recursively Applied and Projected MUltiple SIgnal Classification
+    (RAP-MUSIC) on evoked data.
+
+    Parameters
+    ----------
+    evoked : Evoked
+        Evoked data to localize.
+    forward : instance of Forward
+        Forward operator.
+    noise_cov : instance of Covariance
+        The noise covariance.
+    signal_ndim : int
+        The dimension of the subspace spanning the signal.
+        The default value is the number of dipoles.
```

Comment:
if signal_ndim depends on n_dipoles params I would put n_dipoles first in the list of params